packages/compass-saved-aggregations-queries/src/hooks/use-grid-filters.tsx

```diff
@@ -0,0 +1,278 @@
+  marginRight: spacing[2],
+});
+
+const filterStyles = css({
+  display: 'flex',
+});
+
+const searchInputStyles = css({
+  width: '300px',
+  marginRight: spacing[2],
+});
+
+const RenderFilterSelect = ({
+  items,
+  onSelect,
+  defaultValue,
+  placeHolder = 'Select',
+}: {
+  items: Pick<Tree, 'value'>[];
+  onSelect: (value: string) => void;
+  defaultValue?: string;
+  placeHolder?: string;
+}) => {
+  const labelId = useId();
+  const controlId = useId();
+  const longestLabel = Math.max(
+    placeHolder.length,
+    ...items.map((item) => item.value.length)
+  );
+  return (
+    <Select
+      disabled={items.length === 0}
+      id={controlId}
+      aria-labelledby={labelId}
+      allowDeselect={true}
+      placeholder={placeHolder}
+      className={cx(
+        selectStyles,
+        css({ minWidth: `calc(${longestLabel}ch + ${spacing[6]}px)` })
+      )}
+      onChange={onSelect}
+      defaultValue={defaultValue}
+    >
+      {items.map((item) => (
+        <Option key={item.value} value={item.value}>
+          {item.value}
+        </Option>
+      ))}
+    </Select>
+  );
+};
+
+function useSearchFilter(): [React.ReactElement, string] {
+  const ref = useRef<HTMLInputElement>(null);
+  const [search, setSearch] = useState('');
+  useEffect(() => {
+    ref.current?.focus();
+  }, [search]);
+  const searchControls = useMemo(() => {
+    return (
+      <TextInput
+        ref={ref}
+        className={searchInputStyles}
+        aria-label="Search"
+        type="search"
+        placeholder="Search"
+        value={search}
+        defaultValue={search}
+        onChange={(e) => {
+          setSearch(e.target.value);
+        }}
+        spellCheck={false}
+      />
+    );
+  }, [search]);
+
+  return [searchControls, search];
+}
+
+function useSelectFilter(tree: Tree[]): [React.ReactElement, SelectState] {
+  const [selectState, setSelectState] = useState<SelectState>({});
+  const [collections, setCollections] = useState<Tree[]>([]);
+
+  const selectDatabase = useMemo(() => {
+    return (database: string): void => {
+      setSelectState({
+        database: database ?? undefined,
+      });
+      const items = tree.find((x) => x.value === database)?.items;
+      setCollections(items || []);
+    };
+  }, [tree]);
+
+  const selectCollection = useMemo(() => {
+    return (collection: string): void => {
+      setSelectState({
+        ...selectState,
+        collection: collection ?? undefined,
+      });
+    };
+  }, [selectState]);
+
+  const treeControls = useMemo(() => {
+    return (
+      <div className={selectContainer}>
+        <RenderFilterSelect
+          items={tree}
+          placeHolder={'All databases'}
+          onSelect={selectDatabase}
+          defaultValue={selectState.database}
+        />
+        <RenderFilterSelect
+          items={collections}
+          placeHolder={'All collections'}
+          onSelect={selectCollection}
+          defaultValue={selectState.collection}
```

Comment:
As you have this in your local state you probably want them to always be in sync

```suggestion
          value={selectState.database}
        />
        <RenderFilterSelect
          items={collections}
          placeHolder={'All collections'}
          onSelect={selectCollection}
          value={selectState.collection}
```